server/plugin/http.go

```
@@ -0,0 +1,112 @@
+	contentType := "application/json"
+	buf := &bytes.Buffer{}
+	if err = json.NewEncoder(buf).Encode(in); err != nil {
+		return nil, http.StatusInternalServerError, err
+	}
+	return c.Call(method, path, contentType, buf, out, params)
+}
+
+func (c *client) Call(method, path, contentType string, inBody io.Reader, out interface{}, params url.Values) (responseData []byte, statusCode int, err error) {
+	errContext := fmt.Sprintf("serviceNow: Call failed: method:%s, path:%s", method, path)
+	pathURL, err := url.Parse(path)
+	if err != nil {
+		return nil, http.StatusInternalServerError, errors.WithMessage(err, errContext)
+	}
+
+	if pathURL.Scheme == "" || pathURL.Host == "" {
+		var baseURL *url.URL
+		baseURL, err = url.Parse(c.plugin.getConfiguration().ServiceNowBaseURL)
+		if err != nil {
+			return nil, http.StatusInternalServerError, errors.WithMessage(err, errContext)
+		}
+
+		if path[0] != '/' {
+			path = "/" + path
+		}
+		path = baseURL.String() + path
+	}
+
+	req, err := http.NewRequest(method, path, inBody)
+	if err != nil {
+		return nil, http.StatusInternalServerError, err
+	}
+	if params != nil {
+		req.URL.RawQuery = params.Encode()
+	}
+	if contentType != "" {
+		req.Header.Add("Content-Type", contentType)
+	}
+
+	resp, err := c.httpClient.Do(req)
+	if err != nil {
+		if strings.Contains(err.Error(), "invalid character '<'") {
+			return nil, http.StatusInternalServerError, ErrorContentTypeNotJSON
+		}
+
+		updatedError := strings.ReplaceAll(err.Error(), c.plugin.getConfiguration().ServiceNowBaseURL, "")
+		if strings.Contains(err.Error(), "dial tcp") {
+			errorData := strings.Split(c.plugin.getConfiguration().ServiceNowBaseURL, ":")
+			if len(errorData) == 3 {
+				updatedError = strings.ReplaceAll(updatedError, errorData[2], "")
+			}
+		}
+
+		return nil, http.StatusInternalServerError, errors.New(updatedError)
+	}
```

The current approach to fix the issue #148 is incorrect as it just looks for a specific pattern in the error message and replaces few strings. This might not suppress all the correct errors when an internal service is indeed running and that service throws a different error. The recommended way to fix this issue is to handle the http connection error and return a standard error message indicating "Unable to make connection to the specified ServiceNow instance". We should not return back any other error messages. For debugging purpose (if required), you can print the `err.Error()` in the server logs.

---

```
if err != nil {
		if strings.Contains(err.Error(), "invalid character '<'") {
			return nil, http.StatusInternalServerError, ErrorContentTypeNotJSON
		}

		if strings.Contains(err.Error(), "connect: connection refused") {
			return nil, http.StatusInternalServerError, ErrorConnectionRefused
		}

		return nil, http.StatusInternalServerError, err
	}

```

@srkgupta are you suggesting something like this? Here we are checking the connection error on the basis of whether it contains `connect: connection refused` words in it and returning a generic error to handle the http connection error, otherwise we return an error as it is.

---

Hi @ayusht2810. No, we should not check for any string patterns. We should just return a standard error message if `err != nil`

We can print the err.Error() as log statements. But we should not return the actual error message as server response.